cmd/ocsp-responder/main.go

```diff
@@ -190,25 +210,163 @@ func (src *dbSource) Response(ctx context.Context, req *ocsp.Request) ([]byte, h
+		}
+		// Parse the OCSP bytes returned from the primary source to check
+		// status, expiration and other fields.
+		primaryParsed, err := ocsp.ParseResponse(primaryResult.bytes, nil)
+		if err != nil {
+			src.log.AuditErrf("parsing OCSP response: %s", err)
+			ocspLookups.WithLabelValues("mysql_failed").Inc()
+			sourceUsed.WithLabelValues("error_returned").Inc()
+			return nil, nil, err
+		}
+
+		secondaryParsed, err := ocsp.ParseResponse(secondaryResult.bytes, nil)
+		if err != nil {
+			src.log.Debugf("secondary OCSP lookup response error: %v", err)
+			ocspLookups.WithLabelValues("redis_failed").Inc()
+			sourceUsed.WithLabelValues("mysql").Inc()
+			return primaryResult.bytes, header, nil
+		}
+		if primaryParsed.Status != secondaryParsed.Status {
+			src.log.Err("primary ocsp source doesn't match secondary source, returning primary response")
+			ocspLookups.WithLabelValues("redis_mismatch").Inc()
+			sourceUsed.WithLabelValues("mysql").Inc()
+			return primaryResult.bytes, header, nil
+		}
+		src.log.Debugf("returning ocsp from secondary source: %v", helper.PrettyResponse(secondaryParsed))
+		ocspLookups.WithLabelValues("redis_success").Inc()
+		sourceUsed.WithLabelValues("redis").Inc()
+		return secondaryResult.bytes, header, nil
 	}
-	return certStatus.OCSPResponse, header, nil
+}
+
+type ocspLookup interface {
+	getResponse(context.Context, *ocsp.Request) chan lookupResponse
+}
+
+type redisReceiver struct {
+	rocspReader *rocsp.Client
+}
+type dbReceiver struct {
+	dbMap  dbSelector
+	filter *ocspFilter
+	log    blog.Logger
+}
+
+type lookupResponse struct {
+	bytes []byte
+	err   error
+}
+
+func (src dbReceiver) getResponse(ctx context.Context, req *ocsp.Request) chan lookupResponse {
+	responseChan := make(chan lookupResponse)
+	serialString := core.SerialToString(req.SerialNumber)
+
+	go func() {
+		defer close(responseChan)
+		certStatus, err := sa.SelectCertificateStatus(src.dbMap.WithContext(ctx), serialString)
+		if err != nil {
+			if db.IsNoRows(err) {
+				responseChan <- lookupResponse{nil, bocsp.ErrNotFound}
+				return
+			}
+			responseChan <- lookupResponse{nil, err}
+		}
+
+		if certStatus.IsExpired {
+			src.log.Infof("OCSP Response not sent (expired) for CA=%s, Serial=%s", hex.EncodeToString(req.IssuerKeyHash), serialString)
+			responseChan <- lookupResponse{nil, bocsp.ErrNotFound}
+			return
+		} else if certStatus.OCSPLastUpdated.IsZero() {
+			src.log.Warningf("OCSP Response not sent (ocspLastUpdated is zero) for CA=%s, Serial=%s", hex.EncodeToString(req.IssuerKeyHash), serialString)
+			responseChan <- lookupResponse{nil, bocsp.ErrNotFound}
+			return
+		} else if !src.filter.responseMatchesIssuer(req, certStatus) {
+			src.log.Warningf("OCSP Response not sent (issuer and serial mismatch) for CA=%s, Serial=%s", hex.EncodeToString(req.IssuerKeyHash), serialString)
+			responseChan <- lookupResponse{nil, bocsp.ErrNotFound}
+			return
+		}
+		responseChan <- lookupResponse{certStatus.OCSPResponse, err}
+
```

Comment:
Can you add an explanation in the PR description about why the DB path has a bunch of filter checks and the Redis path doesn't? Are there equivalent checks on the Redis side we should be doing? I think not; these are some consistency checks that only make sense in the certificateStatus table, which is more detailed, but worth thinking through and writing down.

Reply:
Updated the PR description.